cmd/blockhashes.cpp

```diff
@@ -38,102 +39,12 @@ int main(int argc, char* argv[]) {
-                // Should not happen but you never know
-                throw std::runtime_error("Bad headers sequence. Expected " + std::to_string(expected_block_number) +
-                                         " got " + std::to_string(reached_block_number));
-            }
-
-            // We reached a valid block height in proper sequence
-            // Load data into collector
-            Bytes mdb_data_as_bytes{db::from_mdb_val(mdb_data)};
-            etl::Entry etl_entry{/* hash */ mdb_data_as_bytes, /* block number */ mdb_key_as_bytes};
-            collector.collect(etl_entry);
-
-            // Save last processed block_number and expect next in sequence
-            ++blocks_processed_count;
-            block_number = expected_block_number++;
-            rc = canonical_hashes_table->get_next(&mdb_key, &mdb_data);
-        }
-
-        if (rc && rc != MDB_NOTFOUND) { /* MDB_NOTFOUND is not actually an error rather eof */
-            lmdb::err_handler(rc);
-        }
-
-        SILKWORM_LOG(LogLevel::Info) << "Entries Collected << " << blocks_processed_count << std::endl;
-
-        // Proceed only if we've done something
-        if (blocks_processed_count) {
-            SILKWORM_LOG(LogLevel::Info) << "Started BlockHashes Loading" << std::endl;
-
-            /*
-             * If we're on first sync then we shouldn't have any records in target
-             * table. For this reason we can apply MDB_APPEND to load as
-             * collector (with no transform) ensures collected entries
-             * are already sorted. If instead target table contains already
-             * some data the only option is to load in upsert mode as we
-             * cannot guarantee keys are sorted amongst different calls
-             * of this stage
-             */
-            auto target_table{txn->open(db::table::kHeaderNumbers, MDB_CREATE)};
-            size_t target_table_rcount{0};
-            lmdb::err_handler(target_table->get_rcount(&target_table_rcount));
-            unsigned int db_flags{target_table_rcount ? 0u : MDB_APPEND};
-
-            // Eventually load collected items with no transform (may throw)
-            collector.load(target_table.get(), nullptr, db_flags, /* log_every_percent = */ 10);
-
-            // Update progress height with last processed block
-            db::stages::set_stage_progress(*txn, db::stages::kBlockHashesKey, block_number);
-            lmdb::err_handler(txn->commit());
-
-        } else {
-            SILKWORM_LOG(LogLevel::Info) << "Nothing to process" << std::endl;
-        }
-
-        SILKWORM_LOG(LogLevel::Info) << "All Done" << std::endl;
+        lmdb::DatabaseConfig db_config{db_path};
+        db_config.set_readonly(false);
+        std::shared_ptr<lmdb::Environment> env{lmdb::get_env(db_config)};
+        std::unique_ptr<lmdb::Transaction> txn{env->begin_ro_transaction()};
+        stagedsync::stage_blockhashes(db_path, txn.get());
```

Comment:
I think every single stage should return some sort of code stating their work has completed correctly or some condition has happened. I see the stage_blockhashes is in a try catch block but this is limited to a proper program flow execution. I mean we don't have info back from the stage whether or not we can recover from some condition.
My two cents.